src/app/interaction-model/tests/TestEventEmitting.cpp

```diff
@@ -0,0 +1,164 @@
+        ReturnErrorOnFailure(writer.EndContainer(outerType));
+        ReturnErrorOnFailure(writer.Finalize());
+        mLastEncodedSpan = ByteSpan(mLastEventEncodeBuffer, writer.GetLengthWritten());
+
+        mLastOptions         = options;
+        generatedEventNumber = ++mCurrentEventNumber;
+
+        return CHIP_NO_ERROR;
+    }
+
+    EventNumber CurrentEventNumber() const { return mCurrentEventNumber; }
+    const EventOptions & LastOptions() const { return mLastOptions; }
+    ByteSpan LastWrittenEvent() const { return mLastEncodedSpan; }
+
+    // This relies on the default encoding of events which uses
+    // DataModel::Encode on a EventDataIB::Tag::kData
+    template <typename T>
+    CHIP_ERROR DecodeLastEvent(T & dest)
+    {
+        // attempt to decode the last encoded event
+        TLV::TLVReader reader;
+        TLV::TLVType outerType;
+
+        reader.Init(LastWrittenEvent());
+
+        ReturnErrorOnFailure(reader.Next());
+        ReturnErrorOnFailure(reader.EnterContainer(outerType));
+
+        ReturnErrorOnFailure(reader.Next()); // MUST be positioned on the first element
+        ReturnErrorOnFailure(DataModel::Decode(reader, dest));
+
+        ReturnErrorOnFailure(reader.ExitContainer(outerType));
+
+        return CHIP_NO_ERROR;
+    }
+
+private:
+    EventNumber mCurrentEventNumber = 0;
+    EventOptions mLastOptions;
+    uint8_t mLastEventEncodeBuffer[128];
+    ByteSpan mLastEncodedSpan;
+};
+
+} // namespace
+
+TEST(TestInteractionModelEventEmitting, TestBasicType)
+{
+    LogOnlyEvents logOnlyEvents;
+    Events * events = &logOnlyEvents;
+
+    StartUpEventType event{ kFakeSoftwareVersion };
+
+    EventNumber n1 = events->GenerateEvent(event, 0 /* EndpointId */);
+    ASSERT_EQ(n1, logOnlyEvents.CurrentEventNumber());
+    ASSERT_EQ(logOnlyEvents.LastOptions().mPath,
+              ConcreteEventPath(0 /* endpointId */, StartUpEventType::GetClusterId(), StartUpEventType::GetEventId()));
+
+    chip::app::Clusters::BasicInformation::Events::StartUp::DecodableType decoded_event;
+    CHIP_ERROR err = logOnlyEvents.DecodeLastEvent(decoded_event);
+
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(EventLogging, "Decoding failed: %" CHIP_ERROR_FORMAT, err.Format());
+    }
+    ASSERT_EQ(err, CHIP_NO_ERROR);
+    ASSERT_EQ(decoded_event.softwareVersion, kFakeSoftwareVersion);
+
+    EventNumber n2 = events->GenerateEvent(event, 1);
+    ASSERT_EQ(n2, logOnlyEvents.CurrentEventNumber());
+    ASSERT_NE(n1, logOnlyEvents.CurrentEventNumber());
+
+    ASSERT_EQ(logOnlyEvents.LastOptions().mPath,
+              ConcreteEventPath(1 /* endpointId */, StartUpEventType::GetClusterId(), StartUpEventType::GetEventId()));
+}
+
+TEST(TestInteractionModelEventEmitting, TestFabricScoped)
+{
+    constexpr NodeId kTestNodeId           = 0x12ab;
+    constexpr uint16_t kTestPasscode       = 12345;
+    constexpr FabricIndex kTestFabricIndex = kMinValidFabricIndex + 10;
```

Comment:
nit, why `+ 10`

Reply:
I have no reason for this, did not want to explicitly pick a min or a max. I felt that a constant with `Test` in name already meant `this is a magic number that I just picked` ... I was afraid that min may be too likely to hit special cases. Realistically it does not matter ... I can change it to anything if it seems better.